### PR TITLE
add x86 Intel FSP minimal and experimental support

### PR DESCRIPTION
**Pull request by rizlik**
- tested only on qemu
- boot from the reset vector and then uses FSP binaries to initialize the memory and the silicon
- qemu fsp can be obtained from the edk2 project see https://github.com/wolfSSL/wolfBoot-examples/pull/8 for an example of how to test the PR

- there is also some unrelated bug fixing/improvements

### CODE REVIEW

File: docs/Targets.md [hunk 1]
```diff
@@ -1384,3 +1384,36 @@ make test-sim-internal-flash-with-update
 # it should print 2
 ./wolfboot.elf success get_version
 ```
+## Intel X86 and X86_64 bare metal with Intel FSP support (experimental)
```
Comment (dgarske) [inline docs/Targets.md:1387]: Add two newlines above and add link to item on top of document.

File: hal/x86_fsp.c [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA
+ */
+
+#include <wolfboot/wolfboot.h>
+#include <stdint.h>
+
+#ifdef __WOLFBOOT
+void hal_init(void)
```
Comment (dgarske) [inline hal/x86_fsp.c:26]: All functions in this HAL are stubs?

File: include/x86/fsp.h [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+ * Definition of headers used in Intel FSP binaries
+ */
+
+#include <stdint.h>
+#include <x86/hob.h>
+
+#ifndef FSP_H
+#define FSP_H
+
+struct fsp_info_header {
```
Comment (dgarske) [inline include/x86/fsp.h:30]: Is this a packed struct?

File: include/x86/serial.h [hunk 4]
```diff
@@ -0,0 +1,11 @@
+#ifndef SERIAL_H
+#define SERIAL_H
+
+#include <stdint.h>
+
+#define SERIAL_PORT 0x3f8
+
+int uart_init();
```
Comment (dgarske) [inline include/x86/serial.h:8]: You might consider reusing some of the `DEBUG_UART` stuff. See `include/printf.h`. It already declares these functions and maps them to `wolfBoot_printf`

File: src/x86/fsp/machines/qemu/qemu_fsp.c [hunk 5]
```diff
@@ -0,0 +1,38 @@
+#include <stdint.h>
```
Comment (dgarske) [inline src/x86/fsp/machines/qemu/qemu_fsp.c:1]: Make sure the standard wolfBoot header is added to all new files.

File: src/x86/fsp/machines/qemu/qemu_fsp.c [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+#include <x86/hob.h>
+#include "qemu_fsp.h"
+
+const
+FSPT_UPD TempRamInitParams = {
```
Comment (dgarske) [inline src/x86/fsp/machines/qemu/qemu_fsp.c:6]: Use 4 space indentation.

**Review by dgarske — CHANGES_REQUESTED** [review]
Great work! Haven't been able to test yet. Just a couple of minor cleanups so far.

File: hal/x86_fsp.c (hunk 2 above)
Comment (rizlik) [reply hal/x86_fsp.c:26]: yes. In the future, we may implement flash/pci-e/watchdog here

File: include/x86/fsp.h (hunk 3 above)
Comment (rizlik) [reply include/x86/fsp.h:30]: yes. added it, thanks.

File: include/x86/serial.h (hunk 4 above)
Comment (rizlik) [reply include/x86/serial.h:8]: refactored `serial.c` in `hal/x86_uart.c` to better match wolfboot uart hal

**Review by danielinux — DISMISSED** [review]
Tested via https://github.com/wolfSSL/wolfBoot-examples/pull/8 - works as expected.

**Comment by danielinux** [comment]
Development moved to another WIP branch

**Pull request closed without merging**